### PR TITLE
vim-patch:9.0.0844: handling 'statusline' errors is spread out

### PR DESCRIPTION
**Pull request by luukvbaal**
**vim-patch:9.0.0844: handling 'statusline' errors is spread out**

Problem: Handling 'statusline' errors is spread out.
Solution: Pass the option name to the lower levels so the option can be
                reset there when an error is encountered. (Luuk van Baal,
                closes vim/vim#11467)

https://github.com/vim/vim/commit/7b224fdf4a29f115567d4fc8629c1cef92d8444a

Co-authored-by: Luuk van Baal <luukvbaal@gmail.com>

### CODE REVIEW

File: src/nvim/statusline.c [hunk 1]
```diff
@@ -634,6 +620,10 @@ int build_stl_str_hl(win_T *wp, char *out, size_t outlen, char *fmt, int use_san
   char *usefmt = fmt;
   const int save_must_redraw = must_redraw;
   const int save_redr_type = curwin->w_redr_type;
+  // TODO(Bram): find out why using called_emsg_before makes tests fail, does it
+  // matter?
+  // const int        called_emsg_before = called_emsg;
```
Comment (luukvbaal) [inline src/nvim/statusline.c:625]: Do we include these comments?

**Comment by luukvbaal** [comment]
Wow porting didn't go too well, I missed a a lot😅

File: test/functional/lua/ffi_spec.lua [hunk 2]
```diff
@@ -35,11 +35,12 @@ describe('ffi.cdef', function()
 
         int build_stl_str_hl(
           win_T *wp,
-          char_u *out,
+          char *out,
```
Comment (luukvbaal) [inline test/functional/lua/ffi_spec.lua:38]: `char_u` usage seems outdated here?

File: src/nvim/statusline.c (hunk 1 above)
Comment (zeertzjq) [reply src/nvim/statusline.c:622]:
```suggestion
  // const int called_emsg_before = called_emsg;
```

File: src/nvim/statusline.c [hunk 3]
```diff
@@ -647,6 +634,10 @@ int build_stl_str_hl(win_T *wp, char *out, size_t outlen, char *fmt, int use_san
     stl_separator_locations = xmalloc(sizeof(int) * stl_items_len);
   }
 
+  // if "fmt" was set insecurely it needs to be evaluated in the sandbox
+  const int use_sandbox = opt_name ? was_set_insecurely(wp, opt_name, opt_scope)
```
Comment (luukvbaal) [inline src/nvim/statusline.c:638]: API function `nvim_eval_statusline()` will not pass along an option name. Opted to pass along NULL instead and null check here. This patch didn't end up porting as cleanly as I would have hoped but I guess it's still an improvement.

Comment (zeertzjq) [reply src/nvim/statusline.c:638]: I think you can add a comment here about it.

**Review by zeertzjq — APPROVED** [review]

File: test/functional/lua/ffi_spec.lua [hunk 4]
```diff
@@ -35,11 +35,12 @@ describe('ffi.cdef', function()
...
           size_t outlen,
-          char_u *fmt,
-          int use_sandbox,
-          char_u fillchar,
+          char *fmt,
+          char *opt_name,
+          int opt_scope,
+          char fillchar,
```
Comment (zeertzjq) [inline test/functional/lua/ffi_spec.lua:43]:
```suggestion
          int fillchar,
```

**Pull request merged**